### PR TITLE
Prevent Lua from allocating all heap memory (Horus)

### LINKED ISSUE #4284: Prevent Lua from allocating all heap memory
We need to limit the amount of heap memory that Lua allocates on Horus. 

Currently there is no limit other than the amount of the actual physical memory size. So when bogus Lua script uses up all memory, the rest of the firmware that also uses heap (like model images) will have problems working. The user interface will be affected.

The Taranis is not affected, because it doesn't use heap in other parts of firmware.

Proposed solution:
 * keep track of all heap memory used up by Lua and limit that number to something less (lets call it **system reserve**) than the actual max heap size. 
 * do not forget to count all memory allocations done by Lua scripts that are not counted by the Lua, like bitmaps.

Open questions:
 * how big should system reserve be?
 * what to do to Lua when our calculation shows excessive memory usage. We have two independent Lua states one for widgets/themes other for mix/function/standalone scripts. How to decide which script to kill or even which Lua state. Tracking of Lua memory usage per script is not possible (I think), we can only get the memory usage of the whole Lua state.

### PR DESCRIPTION
**Pull request by projectkk2glider**
Lua bitmap functions cleanup and docs update
Added x or y < 0 check to drawBitmap()

Closes #4284

### CODE REVIEW

File: radio/src/gui/480x272/bitmapbuffer.h [hunk 1]
```diff
@@ -255,9 +255,9 @@ class BitmapBuffer: public BitmapBufferBase<uint16_t>
     template<class T>
     void drawBitmap(coord_t x, coord_t y, const T * bmp, coord_t srcx=0, coord_t srcy=0, coord_t w=0, coord_t h=0, float scale=0)
     {
-      if (!data || !bmp || x >= width || y >= height)
+      if (!data || !bmp || x < 0 || x >= width || y < 0 || y >= height)
```
Comment (projectkk2glider) [inline radio/src/gui/480x272/bitmapbuffer.h]: `coord_t` is signed!

File: radio/src/lua/api_lcd.cpp [hunk 2]
```diff
@@ -392,11 +411,18 @@ static int luaGetBitmapSize(lua_State * L)
 
 static int luaDestroyBitmap(lua_State * L)
 {
-  BitmapBuffer * ptr = checkBitmap(L, 1);
-  uint32_t size = ptr->getDataSize();
-  TRACE("luaDestroyBitmap: %p (%u)", ptr, size);
-  if (luaExtraMemoryUsage > size) luaExtraMemoryUsage -= size;
-  delete ptr;
+  BitmapBuffer * b = checkBitmap(L, 1);
+  if (b) {
```
Comment (projectkk2glider) [inline radio/src/lua/api_lcd.cpp]: `b` can be NULL

File: radio/src/lua/api_lcd.cpp [hunk 3]
```diff
@@ -434,15 +461,19 @@ Displays a bitmap at (x,y)
...
-  unsigned scale = luaL_optunsigned(L, 4, 0);
+  const BitmapBuffer * b = checkBitmap(L, 1);
 
-  if (bitmap) {
-    lcd->drawBitmap(x, y, bitmap, 0, 0, 0, 0, (float) scale/100);
+  if (b) {
+    unsigned int x = luaL_checkunsigned(L, 2);
+    unsigned int y = luaL_checkunsigned(L, 3);
+    unsigned int scale = luaL_optunsigned(L, 4, 0);
+    if (scale) {
```
Comment (projectkk2glider) [inline radio/src/lua/api_lcd.cpp]: Small optimization if `scale == 0`, one less floating point division

File: radio/src/lua/interface.cpp [hunk 4]
```diff
@@ -954,6 +954,27 @@ bool luaTask(event_t evt, uint8_t scriptType, bool allowLcdUsage)
...
 
+void checkLuaMemoryUsage()
+{
+#if (LUA_MEM_MAX > 0)
+  uint32_t totalMemUsed = luaGetMemUsed(lsScripts);
+#if defined(COLORLCD)
+  totalMemUsed += luaGetMemUsed(lsWidgets);
+  totalMemUsed += luaExtraMemoryUsage;
+#endif
+  if (totalMemUsed > LUA_MEM_MAX) {
```
Comment (projectkk2glider) [inline radio/src/lua/interface.cpp]: Did not come up with smarter solution, so in this case both Lua states are disabled.

File: radio/src/lua/api_lcd.cpp [hunk 5]
```diff
@@ -434,15 +461,19 @@ Displays a bitmap at (x,y)
...
+      lcd->drawBitmap(x, y, b, 0, 0, 0, 0, scale/100.0);
```
Comment (mpaperno) [inline radio/src/lua/api_lcd.cpp]: `100.0f` maybe?  Don't mean to pick on small stuff, PR looks great.

Comment (projectkk2glider) [inline radio/src/lua/api_lcd.cpp]: No, no, pick away. This is why we have these reviews. Good catch.

File: radio/src/lua/api_lcd.cpp (hunk 2 above)
Comment (mpaperno) [inline radio/src/lua/api_lcd.cpp]: 👍   Assuming valid pointers seems to be a somewhat common issue in the code.

**Comment by bsongis** [comment]
Thanks!

**Pull request merged**